sklearn/feature_selection/_sequential.py

```diff
@@ -0,0 +1,200 @@
+        self.scoring = scoring
+        self.cv = cv
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Learn the features to select.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training vectors.
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+
+        tags = self._get_tags()
+        X, y = self._validate_data(
+            X, y, accept_sparse="csc",
+            ensure_min_features=2,
+            force_all_finite=not tags.get('allow_nan', True),
+            multi_output=True
+        )
+
+        if self.n_features_to_select is None:
+            self.n_features_to_select_ = X.shape[1] // 2
+        else:
+            self.n_features_to_select_ = self.n_features_to_select
+
+        if not 1 <= self.n_features_to_select_ < X.shape[1]:
+            raise ValueError(
+                "n_features_to_select must be in [1, n_features - 1] = "
+                f"[1, {X.shape[1] - 1}]. Got {self.n_features_to_select_}."
+            )
+
+        cloned_estimator = clone(self.estimator)
+
+        # the current mask corresponds to the set of features:
+        # - that we have already *selected* if we do forward selection
+        # - that we have already *excluded* if we do backward selection
+        current_mask = np.zeros(shape=X.shape[1], dtype=bool)
+        n_iterations = (self.n_features_to_select_ if self.forward
+                        else X.shape[1] - self.n_features_to_select_)
+        for _ in range(n_iterations):
+            new_feature_idx = self._get_best_new_feature(cloned_estimator, X,
+                                                         y, current_mask)
+            current_mask[new_feature_idx] = True
+
+        if not self.forward:
+            current_mask = ~current_mask
+        self.support_ = current_mask
+
+        return self
+
+    def _get_best_new_feature(self, estimator, X, y, current_mask):
+        # Return the best new feature to add to the current_mask, i.e. return
+        # the best new feature to add (resp. remove) when doing forward
+        # selection (resp. backward selection)
+        candidate_feature_indices = np.flatnonzero(~current_mask)
+        scores = {}
+        for feature_idx in candidate_feature_indices:
+            candidate_mask = current_mask.copy()
+            candidate_mask[feature_idx] = True
+            if not self.forward:
+                # For backward selection, we transform candidate_mask into its
+                # complement, i.e. we change its semantic from "features to
+                # remove" to "features to keep" because _safe_indexing only
+                # understands the latter
+                # TODO: maybe remove when _safe_indexing supports "complement"
+                candidate_mask = ~candidate_mask
+            X_new = _safe_indexing(X, candidate_mask, axis=1)
```

Comment:
Do we need `_safe_indexing` here?

Reply:
why wouldn't we?

Reply:
My impression is that `_safe_indexing` is used when we want to support `X` as a dataframe. If we are sure it is a ndarray or a sparse matrix, then using plain slicing?

(In it's current state, `_validate_data` will always return a ndarray or a sparse matrix)